source/sdk/lang/Exception.ooc

```diff
@@ -1,397 +1,395 @@
-}
-
-_cleanupStackFrames: inline func {
-  while (!stackFramesToDelete isEmpty) {
-    frame := stackFramesToDelete pop() as StackFrame
-    if (frame)
-      gc_free(frame as Void*)
-  }
-}
-
-_setException: inline func (e: Exception) {
-    _exception set(e)
-}
-
-_getException: inline func -> Exception {
-    _exception get()
-}
-
-_popStackFrame: inline func {
-    frame := exceptionStack get() as Stack<StackFrame> pop() as StackFrame
-    if (frame)
-      gc_free(frame as Void*)
+	new: static func -> This {
+		gc_malloc(_StackFrame size)
+	}
+	free: func {
+		gc_free(this as Void*)
+	}
 }
 
-_takeStackFrame: func -> StackFrame {
-  exceptionStack get() as Stack<StackFrame> pop() as StackFrame
+ExceptionContext: class {
+	_exceptionJumpLabel: Int = 1
+	_localFrames := ThreadLocal<Stack<StackFrame>> new()
+	_currentException := ThreadLocal<Exception> new()
+	_stackFramesToDelete := Stack<StackFrame> new()
+	init: func
+	pushStackFrame: func -> StackFrame {
+		stack: Stack<StackFrame>
+		if (!this _localFrames hasValue?())
+			stack = Stack<StackFrame> new()
+		else
+			stack = this _localFrames get()
+		buf := StackFrame new()
+		stack push(buf)
+		this _localFrames set(stack)
+		buf
+	}
+
+	cleanupStackFrames: func {
+		while (!this _stackFramesToDelete isEmpty) {
+			frame := this _stackFramesToDelete pop() as StackFrame
+			if (frame)
+				frame free()
+		}
+	}
+
+	setException: func (exception: Exception) {
+		this _currentException set(exception)
+	}
+
+	getException: func -> Exception {
+		this _currentException get()
+	}
+
+	popStackFrame: func {
+		frame := this _localFrames get() as Stack<StackFrame> pop() as StackFrame
+		if (frame)
+			frame free()
+		this cleanupStackFrames()
+	}
+
+	takeStackFrame: func -> StackFrame {
+		frame := this _localFrames get() as Stack<StackFrame> pop() as StackFrame
+		this _stackFramesToDelete push(frame)
+		frame
+	}
+
+	hasStackFrame: func -> Bool {
+		this _localFrames hasValue?() && this _localFrames get() as Stack<StackFrame> size > 0
+	}
 }
 
-_hasStackFrame: inline func -> Bool {
-    exceptionStack hasValue?() && exceptionStack get() as Stack<StackFrame> size > 0
-}
+exceptionContext := ExceptionContext new()
 
-assert: extern func(Bool)
+_pushStackFrame: func -> StackFrame { exceptionContext pushStackFrame() }
+_popStackFrame: func -> StackFrame { exceptionContext popStackFrame() }
+_getException: func -> Exception { exceptionContext getException() }
 
 version(windows) {
-    import native/win32/[types, errors]
+	import native/win32/[types, errors]
 
-    getOSErrorCode: func -> Int {
-        GetLastError()
-    }
+	getOSErrorCode: func -> Int {
+		GetLastError()
+	}
 
-    getOSError: func -> String {
-        GetWindowsErrorMessage(GetLastError())
-        
-    }
+	getOSError: func -> String {
+		GetWindowsErrorMessage(GetLastError())
+	}
 } else {
-    errno: extern Int
-    strerror: extern func (Int) -> CString
+	errno: extern Int
+	strerror: extern func (Int) -> CString
```

Comment:
I understand that. I meant non-external functions as well, some return a `String` cast to `CString`.